### DISCUSSION TITLE
Storing application data

### DISCUSSION THREAD

**Discussion by ghost** [tauri-apps/tauri#5557] [Q&A] [answered]
Where should you store application data like language preferences, or other data? 
Is $CONFIG directory the best option for this simple case?

**Comment by FabianLars** [comment, marked as answer]
> Is $CONFIG directory the best option for this simple case?

Yes. We also have this: https://tauri.app/v1/api/js/path/#appdir (also available in rust) to get the path to an app specific directory in the config dir. This is our general recommendation to store _stuff_, but the next tauri release will have similar convenience functions for the other base dirs, for example appDataDir for non-config data

**Reply by escritorio-gustavo** [reply]
How do I get this path from rust? `tauri::api::path::app_local_data_dir` requires a `&Config`, which I don't know how to obtain. I mean, I know I can serialize `tauri.conf.json` into `Config`, but should I? Does tauri make the config object visible anywhere?

**Reply by sjcovell** [reply]
Here's what I've found. Sorry if it's sloppy, I'm still newer to Rust and Tauri.

If you're doing one-time setup stuff then you can do this:
```rust
fn some_setup(app: &mut tauri::App) -> Result<(), Box<dyn std::error::Error>> {
  app.handle().path_resolver().app_data_dir();
  Ok(())
}

fn main() {
  tauri::Builder::default()
    .setup(some_setup)
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}
```

If you're doing the work in a command, then you get an app handler for free:
```rust
#[tauri::command]
fn some_command(app_handle: tauri::AppHandle) -> String {
  let binding = app_handle.path_resolver().app_data_dir().unwrap();
  let app_data_dir = binding.to_str().unwrap();
  
  app_data_dir.to_string()
}
```

Remember to `invoke_handler(tauri::generate_handler![some_command])` in your app init if you do the command route.

**Reply by escritorio-gustavo** [reply]
Awesome! I was doing it in a bunch of lazy_static non-sense, but with this I can swap it for a `tokio::sync::OnceCell` initialized in `setup`

**Reply by itzmesonic** [reply]
How do I do that in v2?

**Reply by FabianLars** [reply]
> How do I do that in v2?

@sjcovell's example should still work, you just have to replace `path_resolver()` with `path()`

If this is not what you're asking for, please expand on what you mean with "that".

**Reply by itzmesonic** [reply]
> > How do I do that in v2?
> 
> @sjcovell's example should still work, you just have to replace `path_resolver()` with `path()`
> 
> If this is not what you're asking for, please expand on what you mean with "that".

And how without rust code??

**Reply by FabianLars** [reply]
The v2 equivalent to the link in my initial message is this: https://v2.tauri.app/reference/javascript/api/namespacepath/#functions

**Comment by dezoito** [comment]
> Here's what I've found. Sorry if it's sloppy, I'm still newer to Rust and Tauri.
> .....
> 
> If you're doing the work in a command, then you get an app handler for free:
> 
> ```rust
> #[tauri::command]
> fn some_command(app_handle: tauri::AppHandle) -> String {
>   let binding = app_handle.path_resolver().app_data_dir().unwrap();
>   let app_data_dir = binding.to_str().unwrap();
>   
>   app_data_dir.to_string()
> }
> ```
> 
Thank you @sjcovell ... that was incredibly helpful.

**Reply by himanshurajora** [reply]
in the latest Tauri versions, it's now:
```rust 
app_handle.path().app_data_dir()
```
